gapis/resolve/dependencygraph/footprint.go

```diff
@@ -0,0 +1,224 @@
+// being executed.
+type DefUseAtom interface {
+	DefUseAtom()
+}
+
+// BackPropagationMachine determines the liveness of Behaviours along the
+// back propagation of the operations recorded in Footprint's Behaviours.
+type BackPropagationMachine interface {
+	// IsAlive check whether a given behaviour, which is specified by its index
+	// and the footprint to which it belongs, should be kept alive. Returns true
+	// if the behaviour should be kept alive, otherwise returns false.
+	IsAlive(behaviourIndex uint64, ft *Footprint) bool
+	// RecordBehaviourEffects records the read/write operations of the given
+	// behaviour specified by its index in the given footprint, and returns the
+	// indices of the alive behaviours that should be kept alive due to recording
+	// of the given behaviour.
+	RecordBehaviourEffects(behaviourIndex uint64, ft *Footprint) []uint64
+	// Clear clears the internal state of the BackPropagationMachine.
+	Clear()
+}
+
+// dummyMachine does nothing but marks all the incoming Behaviours as alive.
+type dummyMachine struct{}
+
+func (m *dummyMachine) IsAlive(uint64, *Footprint) bool { return true }
+func (m *dummyMachine) RecordBehaviourEffects(behaviourIndex uint64,
+	ft *Footprint) []uint64 {
+	return []uint64{behaviourIndex}
+}
+
+func (m *dummyMachine) Clear() {}
+
+// GetBehaviourIndex returns the index of the last Behaviour in the Footprint
+// which belongs to the command or subcomand indexed by the given SubCmdIdx. In
+// case the SubCmdIdx is invalid or a valid Behaviour index is not found, error
+// will be logged and uint64(0) will be returned.
+func (fpt *Footprint) GetBehaviourIndex(ctx context.Context,
+	fci api.SubCmdIdx) uint64 {
+	v := fpt.commandIndexToBehaviourIndexLookupTable.Value(fci)
+	if v != nil {
+		if u, ok := v.(uint64); ok {
+			return u
+		}
+		log.E(ctx, "Invalid behaviour index: %v is not a uint64", v)
+		return uint64(0)
+	}
+	log.E(ctx, "Cannot get behaviour index for command indexed with: %v", fci)
+	return uint64(0)
+}
+
+// AddBehaviour adds the given Behaviour to the Footprint and updates the
+// internal mapping from SubCmdIdx to the last Behaviour that belongs to that
+// command or subcommand.
+func (fpt *Footprint) AddBehaviour(ctx context.Context, b *Behaviour) bool {
+	bi := uint64(len(fpt.Behaviours))
+	fci := b.BelongTo
+	fpt.commandIndexToBehaviourIndexLookupTable.SetValue(fci, bi)
+	fpt.Behaviours = append(fpt.Behaviours, b)
+	fpt.BehaviourIndices[b] = bi
+	return true
+}
+
+// FootprintBuilderProvider provides FootprintBuilder
+type FootprintBuilderProvider interface {
+	FootprintBuilder(context.Context) FootprintBuilder
+}
+
+// FootprintBuilder incrementally builds Footprint one command by one command.
+type FootprintBuilder interface {
+	BuildFootprint(context.Context, *api.State, *Footprint, api.CmdID, api.Cmd)
+}
+
+// GetFootprint returns a pointer to the resolved Footprint.
+func GetFootprint(ctx context.Context) (*Footprint, error) {
```

Comment:
Exception to the `Get` rule here - there's clearly a `Footprint` type declared in this package, so this is fine.